File: Shared/TokenStore.swift
import Foundation
import Security

/// The OAuth token set, as it sits in the keychain.
struct TokenSet: Codable, Sendable {
    var accessToken: String
    var refreshToken: String
    /// Absolute, not the `expires_in` the server sends: the relative figure is only meaningful at
    /// the instant of the response, and this outlives the process that received it.
    var accessExpiresAt: Date

    /// Treated as expired a minute early, so a token that would have died mid-flight is refreshed
    /// before the request rather than retried after it.
    var isAccessTokenFresh: Bool { accessExpiresAt.timeIntervalSinceNow > 60 }
}

/// Reads and writes the token set in the shared keychain access group.
///
/// Shared because the extension does the work and the app does the signing in: the tokens are
/// written by one process and spent by another, and the keychain is the only store both can reach
/// that the rest of the user's software cannot.
enum TokenStore {

    private static let account = "default"

    private static var baseQuery: [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Configuration.keychainService,
            kSecAttrAccount as String: account,
            kSecAttrAccessGroup as String: Configuration.keychainAccessGroup,
            // The modern keychain, which is the only one that honours an access group on macOS
            // without a keychain-sharing prompt on every read from the second process.
            kSecUseDataProtectionKeychain as String: true,
        ]
    }

    static func load() -> TokenSet? {
        var query = baseQuery
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne

        var result: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        guard status == errSecSuccess, let data = result as? Data else {
            // errSecItemNotFound is ordinary (nobody has signed in). Anything else — a missing
            // entitlement above all — is a misconfiguration that would otherwise be invisible.
            if status != errSecItemNotFound {
                Log.auth.error("keychain read failed: \(KeychainError(status: status).localizedDescription, privacy: .public)")
            }
            return nil
        }
        return try? JSONDecoder().decode(TokenSet.self, from: data)
    }

    static func save(_ tokens: TokenSet) throws {
        let data = try JSONEncoder().encode(tokens)

        // Update first, add on miss: SecItemAdd on an existing item is a duplicate error, and the
        // two calls take different dictionaries, so there is no single "upsert" to make of them.
        let updateStatus = SecItemUpdate(
            baseQuery as CFDictionary,
            [kSecValueData as String: data] as CFDictionary
        )
        if updateStatus == errSecSuccess { return }
        guard updateStatus == errSecItemNotFound else { throw KeychainError(status: updateStatus) }

        var insert = baseQuery
        insert[kSecValueData as String] = data
        // The extension runs without a user session behind it, so anything requiring an unlocked
        // login keychain at first-unlock granularity would fail there and only there.
        insert[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlock

        let addStatus = SecItemAdd(insert as CFDictionary, nil)
        guard addStatus == errSecSuccess else { throw KeychainError(status: addStatus) }
    }

    static func clear() {
        SecItemDelete(baseQuery as CFDictionary)
    }
}

struct KeychainError: LocalizedError {
    let status: OSStatus

    var errorDescription: String? {
        let detail = SecCopyErrorMessageString(status, nil) as String? ?? "status \(status)"
        return "Keychain access failed: \(detail)"
    }
}

File: Windows/App/TokenStore.cs
using System.Security.Cryptography;
using System.Text.Json;

namespace HelmsleyDrive.App;

/// <summary>The OAuth token set, as it sits on disk.</summary>
public sealed record TokenSet(string AccessToken, string RefreshToken, DateTimeOffset AccessExpiresAt)
{
    /// <summary>
    /// Absolute, not the <c>expires_in</c> the server sends: the relative figure is only meaningful
    /// at the instant of the response, and this outlives the process that received it. Treated as
    /// expired a minute early, so a token that would have died mid-flight is refreshed before the
    /// request rather than retried after it.
    /// </summary>
    public bool IsAccessTokenFresh => AccessExpiresAt - DateTimeOffset.UtcNow > TimeSpan.FromMinutes(1);
}

/// <summary>
/// Reads and writes the token set under <see cref="Configuration.DataDirectory"/>, sealed with
/// DPAPI to the current user. The Mac counterpart (Mac/Shared/TokenStore.swift) is bigger because
/// two processes share the credential through a keychain access group; here one process is the
/// only reader, and a user-scoped blob in the profile is the whole of it.
/// </summary>
public static class TokenStore
{
    static string TokenPath => Path.Combine(Configuration.DataDirectory, "oauth-tokens.dat");

    /// <summary>
    /// Entropy mixed into the seal. DPAPI at CurrentUser scope is unsealed by anything running as
    /// the user, which is the whole of what it promises; this raises the price of that from
    /// "read the file and call CryptUnprotectData" to "read the binary first". It is not a secret
    /// and cannot be — a shipped app holds none — and changing it costs one sign-in, because an
    /// older blob then reads as unreadable, which is already handled.
    /// </summary>
    static readonly byte[] Entropy = "HelmsleyDrive/oauth-tokens/v1"u8.ToArray();

    public static TokenSet? Load()
    {
        byte[] protectedBytes;
        try { protectedBytes = File.ReadAllBytes(TokenPath); }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            // Missing is the ordinary case. Held open by a scanner, or unreadable, is not — but
            // neither is worth taking the app down for, and this is read on every redraw.
            if (e is not (FileNotFoundException or DirectoryNotFoundException))
                Console.Error.WriteLine($"token store could not be read: {e.Message}");
            return null;
        }

        byte[]? plain = null;
        try
        {
            plain = ProtectedData.Unprotect(protectedBytes, Entropy, DataProtectionScope.CurrentUser);
            return JsonSerializer.Deserialize<TokenSet>(plain);
        }
        catch (Exception e) when (e is CryptographicException or JsonException)
        {
            // Another user's blob, or one sealed by a build that mixed different entropy.
            // Unreadable either way: the state is "nobody has signed in", and the file only says
            // why the next sign-in was needed.
            Console.Error.WriteLine($"token store unreadable: {e.Message}");
            return null;
        }
        finally
        {
            if (plain is not null) CryptographicOperations.ZeroMemory(plain);
        }
    }

    public static void Save(TokenSet tokens)
    {
        Directory.CreateDirectory(Configuration.DataDirectory);
        var plain = JsonSerializer.SerializeToUtf8Bytes(tokens);
        try
        {
            // Staged and swapped, as the snapshots are. The refresh token rotates server-side
            // before this is written, so the copy on its way to disk is the only one left: a
            // truncated file here is not a slow start but a sign-in the user has to do again.
            var staging = TokenPath + ".new";
            File.WriteAllBytes(staging, ProtectedData.Protect(plain, Entropy, DataProtectionScope.CurrentUser));
            File.Move(staging, TokenPath, overwrite: true);
        }
        finally
        {
            CryptographicOperations.ZeroMemory(plain);
        }
    }

    public static void Clear()
    {
        // File.Delete forgives a missing file but not a missing directory, and before the first
        // sign-in there is neither.
        try { File.Delete(TokenPath); File.Delete(TokenPath + ".new"); }
        catch (DirectoryNotFoundException) { }
    }
}
